### ISSUE TITLE
Cross compiling from Linux to Android

### ISSUE THREAD

**Issue by orhun-sketchup** [conan-io/conan-package-tools#99]
I am trying to build an Android project from Linux. I am getting the following message from conan package tools. 
```
Page       :  1
Builds list:
[('settings', {'os': 'Android', 'os.api_level': 24, 'arch': 'x86', 'build_type': 'Debug', 'compiler': 'clang', 'compiler.version': '5.0', 'compiler.libcxx': 'libc++'}), ('options', {}), ('env_vars', {}), ('build_requires', {})]

############## CONAN PACKAGE TOOLS ######################

DEBUG: - Skipped build, compiler mismatch: {'arch': 'x86', 'build_type': 'Debug', 'compiler': 'clang', 'compiler.libcxx': 'libc++', 'compiler.version': '5.0', 'os': 'Android', 'os.api_level': '24'}

#########################################################
```
If I run 
 * `conan create orhun/testing -s os=Android -s os.api_level=24 ...` directly it works.
 * If I remove relevant code from `_run_test_package` in `test_package_runner.py`, it works.

Is the check in `_run_test_package` in error or am I missing something?

Thanks.

**Comment by memsharded** [comment]
Not sure, but it seems that you are using clang and libc++, which from my total ignorance looks strange, I though Android was using gcc compilers.

Have you tried printing ``conan_compiler``, ``conan_compiler_version``, and the values that you are getting from the settings? Thanks for your feedback and reporting!

**Comment by orhun-sketchup** [comment]
1. As of r15c gcc is no longer supported. From [Release Notes](https://developer.android.com/ndk/downloads/revision_history.html#)

> GCC is no longer supported. It is not removed from NDK yet, but is no longer receiving backports. It cannot be removed until after libc++ stabilizes enough to be the default, as some parts of gnustl are still incompatible with Clang.

2. As expected `conan_compiler` and `conan_compiler_version` doesn't match clang.

> ############## CONAN PACKAGE TOOLS ######################
DEBUG: for gcc 7.2
#########################################################

I think it is getting from my default profile. But I don't think it should. Just several lines below this check, package tools will create a temporary profile anyway and will use that for building the package. My system compiler name, version doesn't need to match whatever my current build is.

How did this work on macOS, where the default profile most likely be clang but android used to be gcc? Even if you used clang for android, compiler name still wouldn't match (i.e. apple-clang vs clang), and most likely the versions wouldn't have either, triggering this message.

Note that if I bypass package tools and use conan directly everything works as expected...
```
> conan create orhun/testing -s os=Android -s -s compiler=clang -s compiler.libcxx=libc++ -s compiler.version=5.0 -s os.api_level=24
icu/59.1@orhun/testing: Exporting package recipe
[snip]
Cross-platform from 'Linux' to 'Android'.
```

**Comment by lasote** [comment]
> everything works as expected...
But how is conan locating your toolchain to cross build to Android? It is truly cross-building without specifying any build_require or environment variable pointing to your toolchain?

I mean, the conan_package_tools, as you saw, in `_run_test_package` there is a compiler check that is skipping the build because the "detected" compiler by conan do not match the specified compiler. It's an old check that I think can be deleted because of the latest package tools improvements, but we should check it carefully before removing it. 

But my real concern is, if you don't specify any environment variable like `CC` or `CXX` or a build_require, how will it locate your toolchain? Are these variables already set in your system?

**Comment by orhun-sketchup** [comment]
> It's an old check that I think can be deleted because of the latest package tools improvements, but we should check it carefully before removing it.

That's why I was thinking I was missing something. I generally work on windows and I didn't hit this check previously because it skips the check on windows. 

I think the check at least should verify we are cross building or not (i.e. check whether `default_profile["os"] == settings.get("os")` ). In my case `default_profile["os"]` is Linux but `settings.get("os")` is Android.

At that point I think conan-package-tools could assume user have set their cross build environment correctly. Basically `_detected_compiler_override` is too early if I am setting CC and CXX through conan package dependencies. 

Note: I just noticed what `_detected_compiler_override` does. If I had noticed it earlier, I would have just set these variables in my 'build.py' for my cross compiles to something else. They don't even have to be correct, since conan will set them for me later, during the actual build. See below.

> But how is conan locating your toolchain to cross build to Android?

I have a ndk-toolchain conan package similar to [conan-android-toolchain](https://github.com/lasote/conan-android-toolchain) that sets env_info.CC and env_info.CXX, and my package requires it in its build_requirements.

in ndk-toolchain:
```
    def package_info(self):
        self.env_info.CC  = os.path.join(self.package_folder, 'toolchain', 'bin', 'clang')
        self.env_info.CXX = os.path.join(self.package_folder, 'toolchain', 'bin', 'clang++')
```

in package that uses ndk-toolchain for cross compile:
```
    def build_requirements(self):
        if self.settings.os == 'Android':
            self.build_requires(f'ndk-toolchain/r15c@{self.user}/{self.channel}')
```
